src/common/PrebufferedStreambuf.h

```diff
@@ -12,27 +13,56 @@
  * common case and overflow to a slower heap-allocated buffer when
  * necessary.
  */
-class PrebufferedStreambuf
-  : public std::basic_streambuf<char, std::basic_string<char>::traits_type>
+
+struct prebuffered_data
 {
+private:
   char *m_buf;
   size_t m_buf_len;
+  char *m_pptr;
   std::string m_overflow;
+public:
+  prebuffered_data(char* buf, size_t buf_len)
+  : m_buf(buf), m_buf_len(buf_len), m_pptr(nullptr) {}
+
+  /// return a string copy (inefficiently)
+  std::string get_str() const;
+
+  // returns current size of content
+  size_t size() const;
+
+  // extracts up to avail chars of content
+  int snprintf(char* dst, size_t avail) const;
+  friend class PrebufferedStreambuf;
+};
+
+class PrebufferedStreambuf
+  : public std::basic_streambuf<char, std::basic_string<char>::traits_type>
+{
+  prebuffered_data* m_data;
+  std::ostream os;
```

Comment:
It really seems wierd that streambuf specialization has and ostream.
I did it just to avoid specialization of ostream, just to have streambuf in it.

Reusing ostream and streambuf between logs is the gist of this change.

Reply:
got it. so the ostream and streambuf will both be thread_local and have the same lifetime 👍